### ISSUE TITLE
Is there share memory functions to get display canvas data like xcb_shm?

### ISSUE THREAD

**Issue by guodong** [sidorares/node-x11#100]
Is there share memory functions to get display canvas data like xcb_shm?

**Comment by sidorares** [comment]
no, shm extensions is not implemnted yet. It's relatively straightforward at protocol level, but I need to check what is the status of shm addons for node

**Comment by sidorares** [comment]
https://www.npmjs.com/browse/keyword/shm

**Comment by sidorares** [comment]
http://www.x.org/releases/X11R7.7/doc/xextproto/shm.txt

**Comment by sidorares** [comment]
DRI might be more useful - it serves similar purpose: you ask server to allocate region of video memory for you and then write to it directly bypassing server and serialisation ( similar to shm works only locally of course ) - also on my TODO list - http://www.x.org/releases/X11R7.7/doc/dri2proto/dri2proto.txt

**Comment by guodong** [comment]
Thanks for your reply, and I will check DRI and node shm later. In my application, I must get display flow with low latency, and send it to another process to compress and publish. I think I could add some code to Xorg to implement mmap file to work with node, this may be a just passable way.

**Comment by sidorares** [comment]
what are you displaying? One easy option would be to wrap external application, see example in https://github.com/sidorares/node-x11/blob/master/examples/simple/embed.js For example you can embed mplayer via `wid [parent id]` command line para and stream some dynamic data to it

**Comment by guodong** [comment]
I want to render x11 window to a client browser, just like x11 architecture but use browser as xserver, but not total same as x11 arch, let me use a picture to explain.
![image](https://cloud.githubusercontent.com/assets/858637/9435975/c60ddeac-4a7e-11e5-87d0-937c3f12e252.png)
In one word, I want to make a software which uses browser as display and uses browser tab as window manager.
I need to send mouse and keyboard event to app server and the proxy is writen by node(in order to fit with socket.io in websocket server). Now the xserver changed bitmap is send by another domain sock to proxy, and proxy compress it and send it to websocket server. In this issue, I think if node-x11 has shm functions, there is no need to use another domain socket and reduce the complexity of it.
I am  a totally newbie of x11 and node, and say sorry if my question is naive in this area. So do you have any advices or suggestions on it?

**Comment by sidorares** [comment]
similar to https://github.com/ttaubert/x-server-js or https://github.com/GothAck/javascript-x-server ?

**Comment by sidorares** [comment]
not sure how shm could help you as you going to serialise communication over websockets

**Comment by guodong** [comment]
Thank you, I have already study the 2 projects above, the difference is that I use original xorg as xserver, but not rewrite a new xserver in js, so I need to get rendered bitmap from xorg.

**Comment by sidorares** [comment]
so it's more like RDP or VNC viewer in the browser?

**Comment by guodong** [comment]
Yes, just like VNC, but only render one application window nor the desktop.

**Comment by sidorares** [comment]
"-id windowid" in http://www.karlrunge.com/x11vnc/x11vnc_opts.html

**Comment by guodong** [comment]
really helpful! I'll look into it, thanks again!

**Comment by polpo** [comment]
Sounds like VNC may be the best option for you, but I've had to solve a similar problem and looked a bit into Crouton, which is the project for ChromeOS that lets you run X11. It packs up X11 to run over a websocket. Unfortunately the code for it is very ChromeOS-specific, but it may be something to look into.

**Comment by sidorares** [comment]
@polpo very interesting

**Comment by sidorares** [comment]
The MIT-SHM extension is implemented now (landed in #217, [lib/ext/shm.js](https://github.com/sidorares/node-x11/blob/master/lib/ext/shm.js)): `QueryVersion`, `Attach`, `Detach`, `PutImage`, `GetImage`, `CreatePixmap` and `AttachFd`.

One caveat: node-x11 speaks only the protocol and has no runtime dependencies, and Node.js itself can't create or mmap SysV shared-memory segments — so to actually read/write the segment you need to pair it with a native module (e.g. `shm-typed-array`) that gives you the `shmid` and a view over the memory.

For the original use case (streaming window contents to a browser) VNC per-window as discussed above is still probably the pragmatic answer.
